rfcs/0005-version-matrix.md

```diff
@@ -0,0 +1,477 @@
+
+Not everyone can or will upgrade their Rust toolchain the moment a new version is published.
+This includes potential new contributors, to whom we want to let write their patch without having to start updating their toolchain.
+Beyond that, it will be even more common among application developers who use our libraries.
+It's a question of balance of course, but we want to support toolchains as old as possible.
+
+### We want a widely applicable strategy
+
+In the absence of a well thought out general strategy we will have (potentially ephemeral) ad hoc approaches for every distinct project.
+This means an increased maintenance burden, as problems solved with one approach won't necessarily carry over to another.
+Also there will be more problems to solve, because a mix of strategies will bring us the union of their respective problems.
+
+
+## User-facing explanation
+
+We test and support only the latest versions of all our package dependencies.
+
+Our minimum supported Rust version (MSRV) is largely decided by our dependencies and we just document the fact.
+
+
+## Implementation strategy
+
+### Keep the existing general Linebender CI as a base and extend it
+
+Our current general CI script is the result of years of experience.
+It already has good techniques to avoid some sources of dependency contamination, e.g. feature activations caused by dev dependencies.
+It also already pins the stable Rust toolchain version, allowing us to choose when to update it.
+Those benefits should move forward and as such this RFC doesn't see the need to remove anything and all changes are additions.
+
+### `Cargo.lock` is committed to source control
+
+Having a one true `Cargo.lock` solves a lot of the contributor and CI issues.
+Contributors can easily use the same dependency versions as the CI does.
+The CI won't suddenly start producing errors just because a new dependency version was published that causes issues.
+We get reproducible historic builds.
+We are in control of the timing when dependency updates are introduced.
+
+### `Cargo.lock` is updated regularly
+
+Given that a `Cargo.lock` of a library is used only by the developers of said library, we should not get comfortable with one set of versions for too long.
+`Cargo.lock` should be updated to contain the most recent versions of dependencies on a regular basis.
+Roughly as often as we update the stable Rust toolchain pin in the CI script.
+Waiting longer than that creates too large of a time window for bad user experience issues to remain hidden.
+Updating it too often, e.g. every week just because, will cut down on the benefits of contributor's local environment matching the CI and increase merge conflicts.
+Critically, an update should always happen just before publishing to ensure that we test the unlocked experience that users will have.
+
+### `Cargo.toml` matches `Cargo.lock`
+
+We don't want to claim that our project works with dependency versions that we don't actually test.
```

Comment:
There is experimental cargo support for compiling with minimal library versions. Perhaps we could compile with minimal and latest and assume that everything in between will work. It seems annoying for users if we require a version we don't actually need.

Reply:
As I explain in the section titled *Compiling with minimum dependency versions is possible only when stars align* those experimental Cargo features don't quite work and so testing minimum versions isn't actually possible. Not in the same workspace anyway.

Can you provide some example scenario where a user would want a SemVer compatible lower version beyond the misguided dependency scenario I outlined in the RFC? Especially because they would have to specifically craft their `Cargo.lock` with the `--precise` flag to get such a scenario, which is unlikely to be common.